In [105]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import warnings
warnings.filterwarnings("ignore")

In [107]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [109]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [111]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [115]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
ScaledData = StandardScaler().fit_transform(df_market_data)

In [117]:
# Create a DataFrame with the scaled data
Scaled_transform = pd.DataFrame(ScaledData,columns=['price_change_percentage_24h','price_change_percentage_7d','price_change_percentage_14d','price_change_percentage_30d',	'price_change_percentage_60d','price_change_percentage_200d','price_change_percentage_1y'])

# Copy the crypto names from the original DataFrame
coin = list(df_market_data.index)
coin

# Set the coin_id column as index
Scaled_transform['coin_id'] = coin
Scaled_transform = Scaled_transform.set_index('coin_id')
# Display the scaled DataFrame
Scaled_transform








price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [197]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 12))
k

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [199]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model = KMeans(n_clusters=i, random_state=1)
    k_model.fit(Scaled_transform)
    inertia.append(k_model.inertia_)

In [200]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}


# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)
df_elbow.head()

k     inertia
0  1  287.000000
1  2  212.123342
2  3  165.136752
3  4   79.022435
4  5   66.413051

In [201]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)


:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [208]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)

In [210]:
# Fit the K-Means model using the scaled DataFrame
model.fit(Scaled_transform)

KMeans(n_clusters=4, random_state=1)

In [212]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
k_4 = model.predict(Scaled_transform)

# Print the resulting array of cluster values.
k_4

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 3, 2, 0, 0, 1, 0, 0, 0, 0])

In [214]:
# Create a copy of the scaled DataFrame
Cluster_copied = clusters_predicted = Scaled_transform.copy()

In [216]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
Cluster_copied['predicted_clusters'] = k_4


# Display the copy of the scaled DataFrame
Cluster_copied.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  predicted_clusters  
coin_id                                                       
bitcoin                        -0.251637                   2  
ethereum                       -0.199352                   2  
tether                         -0.282061                   0  
ripple                         -0.295546                   0  
bitcoin-cash                   -0.270317                   2

In [218]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
Cluster_copied.hvplot.scatter(x="price_change_percentage_24h",y="price_change_percentage_7d",by= "predicted_clusters",hover_cols = 'coin_id')

:NdOverlay   [predicted_clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [222]:
# Create a PCA model instance and set `n_components=3`.
pca=PCA(n_components=3)

In [224]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
clustered_pca = pca.fit_transform(Cluster_copied)

# View the scaled PCA data
clustered_pca[:5]

array([[ 0.44890795, -1.24537573, -0.8506404 ],
       [ 0.49536709, -0.899823  , -1.31755943],
       [-0.81884571,  0.07189909,  0.69501507],
       [-0.84035711,  0.08005407,  0.54435952],
       [ 0.81324004, -2.66952035, -1.64332113]])

In [226]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.34871677, 0.31363391, 0.22627118])

In [228]:
sum(pca.explained_variance_ratio_)

0.8886218549859445

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.8886218549859445

In [231]:
# Create a new DataFrame with the PCA data.
clustered_pca_df = pd.DataFrame(
    clustered_pca,
    columns=["PCA1", "PCA2","PCA3"]
)
# Copy the crypto names from the original scaled DataFrame
clustered_pca_df['coin_id'] = list(Cluster_copied.index)

# Set the coin_id column as index
clustered_pca_df =clustered_pca_df.set_index('coin_id')

# Display the scaled PCA DataFrame
clustered_pca_df


#Scaled_transform

PCA1      PCA2      PCA3
coin_id                                           
bitcoin               0.448908 -1.245376 -0.850640
ethereum              0.495367 -0.899823 -1.317559
tether               -0.818846  0.071899  0.695015
ripple               -0.840357  0.080054  0.544360
bitcoin-cash          0.813240 -2.669520 -1.643321
binancecoin           0.822871 -1.682096 -0.905203
chainlink             0.913868 -1.128684 -2.730038
cardano               0.811102 -1.113378 -1.598655
litecoin              0.258728 -0.946884 -1.103560
bitcoin-cash-sv      -1.003921 -0.119410  0.278564
crypto-com-chain     -1.363289  1.245349  0.950712
usd-coin             -0.830643  0.077047  0.707661
eos                  -1.135458  0.171401  0.441569
monero                2.084978 -2.688979 -0.776469
tron                 -0.633198 -0.016911  0.249426
tezos                -0.161908 -0.310904 -1.818521
okb                  -0.996987  1.292574  0.793701
stellar              -0.990512  0.449877  0.052002
cosmos                0.878129 -1.093136 -1.774115
cdai                 -0.874815  0.008933  0.694993
neo                  -1.135199  0.792113  0.499484
wrapped-bitcoin       0.434801 -1.232506 -0.839794
leo-token            -1.110348  0.540593  0.854934
huobi-token          -0.758629  0.094085  0.466658
nem                   0.222636  0.083977  1.584956
binance-usd          -0.826142  0.049070  0.700753
iota                 -1.059844  0.097044 -0.149781
vechain              -1.808829  1.520236  0.527335
zcash                 0.558082 -0.920479 -1.377145
theta-token           1.452903  1.756424  2.506866
dash                 -0.897830  0.134758 -0.197015
ethereum-classic     -0.908899  0.095612  0.194364
ethlend               6.049082  7.033268 -2.235723
maker                 0.307206 -0.565760 -0.968488
havven               -0.507111  2.407910 -0.392697
omisego              -0.236488 -0.071399  1.445063
celsius-degree-token  6.454065 -2.823519  4.810646
ontology             -1.823588  1.561754  0.000663
ftx-token            -0.689158 -0.253900  0.073233
true-usd             -0.823298  0.031016  0.694746
digibyte             -0.770668  0.187671  0.911020

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [239]:
# Create a list with the number of k-values from 1 to 11
K= list(range(1,12))
k

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [247]:
# Create an empty list to store the inertia values
inertia2 = []


# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model = KMeans(n_clusters=i, random_state=1)
    k_model.fit(clustered_pca_df)
    inertia2.append(k_model.inertia_)

In [249]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data2 = {"k": K, "inertia": inertia2}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_2= pd.DataFrame(elbow_data2)
df_elbow_2


k     inertia
0    1  290.622694
1    2  207.307439
2    3  123.126951
3    4   44.130602
4    5   33.217427
5    6   23.374713
6    7   20.943077
7    8   18.666961
8    9   13.640110
9   10   10.287149
10  11    7.958180

In [251]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
df_elbow_2.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)


:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:**4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** no

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [39]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)

In [253]:
# Fit the K-Means model using the PCA data
model.fit(clustered_pca_df)

KMeans(n_clusters=4, random_state=1)

In [255]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
k_4a = model.predict(clustered_pca_df)

# Print the resulting array of cluster values.
k_4a

array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 3, 1, 0, 0, 2, 0, 0, 0, 0])

In [257]:
# Create a copy of the scaled PCA DataFrame
Cluster_copied2 = clustered_pca_df.copy()


# Add a new column to the copy of the PCA DataFrame with the predicted clusters
Cluster_copied2['predicted_clusters'] = k_4a

# Display the copy of the scaled PCA DataFrame
Cluster_copied2 

PCA1      PCA2      PCA3  predicted_clusters
coin_id                                                               
bitcoin               0.448908 -1.245376 -0.850640                   1
ethereum              0.495367 -0.899823 -1.317559                   1
tether               -0.818846  0.071899  0.695015                   0
ripple               -0.840357  0.080054  0.544360                   0
bitcoin-cash          0.813240 -2.669520 -1.643321                   1
binancecoin           0.822871 -1.682096 -0.905203                   1
chainlink             0.913868 -1.128684 -2.730038                   1
cardano               0.811102 -1.113378 -1.598655                   1
litecoin              0.258728 -0.946884 -1.103560                   1
bitcoin-cash-sv      -1.003921 -0.119410  0.278564                   0
crypto-com-chain     -1.363289  1.245349  0.950712                   0
usd-coin             -0.830643  0.077047  0.707661                   0
eos                  -1.135458  0.171401  0.441569                   0
monero                2.084978 -2.688979 -0.776469                   1
tron                 -0.633198 -0.016911  0.249426                   0
tezos                -0.161908 -0.310904 -1.818521                   1
okb                  -0.996987  1.292574  0.793701                   0
stellar              -0.990512  0.449877  0.052002                   0
cosmos                0.878129 -1.093136 -1.774115                   1
cdai                 -0.874815  0.008933  0.694993                   0
neo                  -1.135199  0.792113  0.499484                   0
wrapped-bitcoin       0.434801 -1.232506 -0.839794                   1
leo-token            -1.110348  0.540593  0.854934                   0
huobi-token          -0.758629  0.094085  0.466658                   0
nem                   0.222636  0.083977  1.584956                   0
binance-usd          -0.826142  0.049070  0.700753                   0
iota                 -1.059844  0.097044 -0.149781                   0
vechain              -1.808829  1.520236  0.527335                   0
zcash                 0.558082 -0.920479 -1.377145                   1
theta-token           1.452903  1.756424  2.506866                   0
dash                 -0.897830  0.134758 -0.197015                   0
ethereum-classic     -0.908899  0.095612  0.194364                   0
ethlend               6.049082  7.033268 -2.235723                   3
maker                 0.307206 -0.565760 -0.968488                   1
havven               -0.507111  2.407910 -0.392697                   0
omisego              -0.236488 -0.071399  1.445063                   0
celsius-degree-token  6.454065 -2.823519  4.810646                   2
ontology             -1.823588  1.561754  0.000663                   0
ftx-token            -0.689158 -0.253900  0.073233                   0
true-usd             -0.823298  0.031016  0.694746                   0
digibyte             -0.770668  0.187671  0.911020                   0

In [259]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
Cluster_copied2 .hvplot.scatter(x="PCA1",y="PCA2",by= "predicted_clusters",hover_cols = 'coin_id')

:NdOverlay   [predicted_clusters]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [271]:
# Composite plot to contrast the Elbow curves
# YOUR CODE HERE!
original_curve = df_elbow.hvplot.line(x="k", y="inertia", title="Elbow Curve", xticks=k)
pca_curve = df_elbow_2.hvplot.line(x="k",y="inertia", title="Elbow Curve", xticks=k)


Composite_curve = original_curve + pca_curve
Composite_curve

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [283]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!
st_copied = Cluster_copied.hvplot.scatter(x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="predicted_clusters",
    hover_cols='coin_id'
)
nd_copied = nd_copied = Cluster_copied2.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    by="predicted_clusters",
    hover_cols='coin_id'
)


composite_plot = st_copied + nd_copied
composite_plot

:Layout
   .NdOverlay.I  :NdOverlay   [predicted_clusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [predicted_clusters]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** the PCA data shows tighter clusters among each other vs random placements of cluster 